ion-schema/src/model/constraints/regex.rs

```diff
@@ -0,0 +1,296 @@
+    }
+}
+
+impl Regex {
+    fn pattern(&self) -> &str {
+        self.regex.as_str()
+    }
+}
+
+impl<V: IslVersion> TypeDefinitionBuilder<V> {
+    /// Adds a `regex` constraint to the type being built.
+    ///
+    /// This will panic if the input is invalid. For a safe alternative, see [`Self::try_regex`].
+    pub fn regex(self, regex: impl IonSchemaRegexSource<V>) -> Self {
+        self.with_constraint(Versioned::into_inner(regex.build_regex().unwrap()).into())
+    }
+
+    /// Adds a `regex` constraint to the type being built.
+    pub fn try_regex(self, regex: impl IonSchemaRegexSource<V>) -> IonSchemaResult<Self> {
+        Ok(self.with_constraint(Versioned::into_inner(regex.build_regex()?).into()))
+    }
+}
+
+impl ValidateInternal for Regex {
+    fn validate_internal<'top: 'call, 'call, R>(
+        &'top self,
+        value: &IonSchemaElement<'top>,
+        ctx: &ValidationContext,
+        recorder: &'call mut R,
+    ) -> ControlFlow<()>
+    where
+        R: ViolationRecorder<'top>,
+    {
+        todo!()
+    }
+}
+
+impl<V: IslVersion> WriteAsIsl<V> for Regex
+where
+    TypeArgument: WriteAsIsl<V>,
+{
+    fn write_as_isl<W: ValueWriter>(
+        &self,
+        writer: W,
+        ctx: &WriteContext<V>,
+    ) -> IonSchemaResult<()> {
+        todo!()
+    }
+}
+
+impl<V: IslVersion> ReadConstraint<V> for Regex {
+    fn read_constraint(ion: &Element, ctx: &LoaderContext<V>) -> IonSchemaResult<Option<Self>> {
+        todo!()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::model::constraints::AnyConstraint;
+    use crate::ISL_1_0;
+
+    #[test]
+    fn test_builder() {
+        let type_ = TypeDefinitionBuilder::<ISL_1_0>::new().regex("abc").build();
+
+        assert_eq!(
+            type_.constraints().cloned().collect::<Vec<_>>(),
+            vec![AnyConstraint::Regex(Regex {
+                regex: regex::Regex::new("abc").unwrap(),
+                multiline: false,
+                case_insensitive: false,
+            })]
+        );
+
+        let my_regex = IonSchemaRegexBuilder::new("abc")
+            .case_insensitive(true)
+            .multiline(true)
+            .build()
+            .unwrap();
+        let type_ = TypeDefinitionBuilder::<ISL_1_0>::new()
+            .regex(my_regex)
+            .build();
+
+        assert_eq!(
+            type_.constraints().cloned().collect::<Vec<_>>(),
+            vec![AnyConstraint::Regex(Regex {
+                regex: RegexBuilder::new("abc")
+                    .case_insensitive(true)
+                    .multi_line(true)
+                    .build()
+                    .unwrap(),
+                multiline: true,
+                case_insensitive: true,
+            })]
+        );
+
+        let type_ = TypeDefinitionBuilder::<ISL_1_0>::new()
+            .regex(
+                "abc"
+                    .to_regex_builder()
+                    .case_insensitive(true)
+                    .multiline(true),
```

Comment:
Not important: you could have `IonSchemaRegexSource` provide its own `case_insensitive` and `multiline` methods that were implemented by calling `.to_regex_builder().METHOD()`. Then you could do:

```rust
.regex("abc".case_insensitive(true).multiline(true))
```

Reply:
I like this.